### PR TITLE
rbd: utilize cookie support from rbd for nbd

### PR DESCRIPTION
**Pull request by pkalever**
# Describe what this PR does #

**Problem:**
On remap/attach of device (i.e. nodeplugin restart), there is no way for rbd-nbd to defend if the backend storage is matching with the initial backend storage.

Say, if an initial map request for backend "pool1/image1" got mapped to /dev/nbd0 and the userspace process is terminated (on nodeplugin restart). A next remap/attach (nodeplugin start) request within reattach-timeout is allowed to use /dev/nbd0 for a different backend "pool1/image2"

For example, an operation like below could be dangerous:
```
$ sudo rbd-nbd map --try-netlink rbd-pool/ext4-image
/dev/nbd0
$ sudo blkid /dev/nbd0
/dev/nbd0: UUID="bfc444b4-64b1-418f-8b36-6e0d170cfc04" TYPE="ext4"
$ sudo pkill -15 rbd-nbd   <-- nodeplugin terminate
$ sudo rbd-nbd attach --try-netlink --device /dev/nbd0 rbd-pool/xfs-image
/dev/nbd0
$ sudo blkid /dev/nbd0
/dev/nbd0: UUID="d29bf343-6570-4069-a9ea-2fa156ced908" TYPE="xfs"
```

**Solution:**
rbd-nbd/kernel now provides a way to keep some metadata in sysfs to identify between the device and the backend, so that when a remap/attach request is made, rbd-nbd can compare and avoid such dangerous operations.

With the provided solution, as part of the initial map request, backend cookie (ceph-csi VOLID) can be stored in the sysfs per device config, so that on a remap/attach request rbd-nbd will check and validate if the backend per device cookie matches with the initial map backend with the help of cookie.

At Ceph-csi we use VOLID as device cookie, which will be unique, we pass the VOLID as cookie at map and use the same at the time of attach, that way rbd-nbd can identify backends and their matching devices.

**Requires:**
https://github.com/ceph/ceph/pull/41323
https://lkml.org/lkml/2021/4/29/274

- [x] Require PR https://github.com/ceph/ceph/pull/41323

Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

**Comment by pkalever** [comment]
`DNM` waiting for https://github.com/ceph/ceph/pull/41323

File: internal/util/util.go
Comment (nixpanic) [inline internal/util/util.go]: Please add a unit test for this function in `internal/util/util_test.go`

Comment (nixpanic) [reply internal/util/util.go]: Also, the commit message mentions "package versions"... what packages? Are these RPM, Golang or other versions?

File: internal/rbd/rbd_attach.go
Comment (nixpanic) [inline internal/rbd/rbd_attach.go]: In what case would it not be possible to get the version? If there is no known case, just return the version or an error. No need for returning an additional bool.

You could split this change from the cookie modification, and have a separate commit for it.

File: internal/rbd/rbd_attach.go
Comment (nixpanic) [inline internal/rbd/rbd_attach.go]: move this to a helper function, like `cmdArgs = appendNbdCookieOption(cmdArgs, cookie)` or such

**Review by nixpanic — CHANGES_REQUESTED** [review]
Looks like a good change to have. Left a few minor comments.

Have you considered inspecting the output of `rbd-nbd --help` to see if there is support for :cookie: ? That would make it possible to get support enabled in case some Ceph vendor backports the feature.

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: any specific reason to have this check? instead of that we split stdout and check its length to make sure it's>3?

File: internal/util/util.go
Comment (Madhu-1) [inline internal/util/util.go]: `return !v2.LessThan(v1)` ? or `!(v2.LessThan(v1)==true`?

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: check version only one and store it globally instead of checking it always?

File: internal/util/util.go
Comment (Madhu-1) [inline internal/util/util.go]: are we discarding the error here?

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: instead of depending on the version is it possible to check cookie is present in some help text of the CLI command?

Comment (pkalever) [reply internal/rbd/rbd_attach.go]: I was trying to address this in a more generic way, considering we might have to use the same utilities for other options in the future, today we are checking for `cookie` tomorrow we might want to check for other options or features.

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: This was there already and I didn't want to break it, hence continued. I don't think there would be a possible case where the version cannot be returned.

Sure I will consider making the requested changes in the next respin.

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: As also suggested by Niels at `checkRbdNbdTools`, let's also safely ignore a case that doesn't return the version?

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Ok

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Aren't we doing the same now? `rbdNbdVersion` is collected once and used everytime.

File: internal/util/util.go
Comment (pkalever) [reply internal/util/util.go]: https://github.com/hashicorp/go-version#version-parsing-and-comparison

Should I rather stick to the binary version rather than the package version?

File: internal/util/util.go
Comment (pkalever) [reply internal/util/util.go]: Yeah, I don't know how useful it would be to handle them. Not seen anything in examples handling them gracefully:
https://github.com/hashicorp/go-version#version-parsing-and-comparison

I can add one if you want.

File: internal/util/util.go
Comment (pkalever) [reply internal/util/util.go]: Yes, looks simple I can pick one.

**Review by pkalever — COMMENTED** [review]
@nixpanic @Madhu-1 I think I was on the other side of the coin thinking about reusing the `rbd-nbd --version` for future options instead of looking for `--cookie` or some other option individually, which could be a per option exec.

Also, the utility function introduced `IsVersionSatisfied` is generic enough for reusability.

That said, I would still like to know your preference among `rbd-nbd --version` or `rbd-nbd --help | grep cookie`


Thanks!

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: OH! Do you mean the comparison part as well?

Like move this part to global:

```
isNbdCookie = util.IsVersionSatisfied(rbdNbdCookieSupportVersion, rbdNbdVersion)

```

File: internal/util/util.go
Comment (pkalever) [reply internal/util/util.go]: Done now, changed it to "util: provide utility to parse and compare versions"

File: internal/util/util.go
Comment (pkalever) [reply internal/util/util.go]: done will respin

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: after the recent changes, looks like this is not needed for me.

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [reply internal/rbd/rbd_attach.go]: Won't this become problematic if someone backports the fix to another version? something like a downstream version?

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: this can cause panic if the ver length is not 3? need a  length check?

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: we no longer default to krbd if nbd is not present?

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Upstream: I think this will be considered as a feature, and hence I don't think ceph will backport it to minor z-stream releases.

Downstream: In, downstream we should change this to something else ?

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: oops! right, I thought I added one based on your previous comment, will fix it.

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [reply internal/rbd/rbd_attach.go]: it good to reuse the same code for both upstream and downstream instead of having downstream only change and I have seen few important bug fixes are backported from octopus to nautilus etc. and also if someone is maintaining their own ceph downstream it might break. just trying to cover all cases here instead of depending on the version?

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Not sure on this part, but how does it matter for us here (just curious)?

Comment (Madhu-1) [reply internal/rbd/rbd_attach.go]: if the nbd module was not available on the node we use krbd to map the image by default with this change we don't default to anything we always try to use nbd if it's specified in storageclass?

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: @nixpanic @Madhu-1 I understand your concern, and it kind of makes sense to me in a different angle, I was just thinking on the reusability part.

Okay! I will grep for `cookie` for now to satisfies your review :-)

BTW, do we want to keep the `IsVersionSatisfied` util func as I have also added `TestIsVersionSatisfied` may be for later use or just discard that work? (I'm just trying to save some future cycles)

Thanks!

Comment (Madhu-1) [reply internal/rbd/rbd_attach.go]: Thank you :+1:  if it's not used anywhere IMO it should be removed.

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: > by default with this change we don't default to anything we always try to use nbd if it's specified in storageclass?

I'm sorry, I might be missing something, which part of the code changes from this PR is leading to the behaviour.

Comment (Madhu-1) [reply internal/rbd/rbd_attach.go]: `if volOptions.Mounter == rbdTonbd && hasNBD {` to `if volOptions.Mounter == rbdTonbd`

Comment (pkalever) [reply internal/rbd/rbd_attach.go]: this is only to avoid checking if rbd-nbd is installed or not (which is redundant) rest remain as it is IMHO.

Comment (nixpanic) [reply internal/rbd/rbd_attach.go]: Previously there NBD was only used when `hasNBD` was set. So, in the case where `checkRbdNbdTools()` returned false, a StorageClass configured to use NBD would fallback to krbd instead. This is a change in behaviour, and we do not know if there are any users that rely on it.

Comment (pkalever) [reply internal/rbd/rbd_attach.go]: right!
so the question is
1. should we have `nbdLoaded` variable that checks if the module is loaded, if not fall back to `krbd` ?
2. Or just fail if the storage class asks for `rbd-nbd` and `lsmod` doesn't list it?

We can still avoid hasNBD checking for rbd-nbd binary availability on the container which is redundant.

Comment (pkalever) [reply internal/rbd/rbd_attach.go]: @nixpanic @Madhu-1 check the latest updates, I have left hasNBD for module loaded case, and added `hasNBDCookieSupport` separately. Hope that will keep the old style of fallback intact.

Thanks!

File: internal/rbd/rbd_attach.go
Comment (nixpanic) [inline internal/rbd/rbd_attach.go]: Default is `false`, this temporary changes it to `true`, after which it may become `false` again. That isn't very clean. Better only set it once to `true` after successful detection.

File: internal/rbd/rbd_attach.go
Comment (nixpanic) [inline internal/rbd/rbd_attach.go]: use different formatting please:
```go
var (
    hasNBD = false
    hasNBDCookieSupport = false
)
```

File: internal/rbd/rbd_attach.go
Comment (nixpanic) [inline internal/rbd/rbd_attach.go]: instead of repeating this if statement, add a helper:
```go
cmdArgs = appendRbdNbdCookie(cmdArgs, cookie)

func appendRbdNbdCookie(...) []string {
    if cookie != "" && hasNBDCookieSupport {
        cmdArgs = ....
    }
    return cmdArgs
}
```

File: internal/rbd/rbd_attach.go
Comment (nixpanic) [inline internal/rbd/rbd_attach.go]: prevent this duplication by adding a `appendRbdNbdCookie()` helper?

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Ok

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: sure

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: I don't see much value @nixpanic 

The syntax of cmdArgs in both cases are different

one is:

```
if cookie != "" && hasNBDCookieSupport {
        cmdArgs = append(cmdArgs, "--options", fmt.Sprintf("cookie=%s", cookie))
}

```

and the other is:
```
if cookie != "" && hasNBDCookieSupport {
        cmdArgs = append(cmdArgs, fmt.Sprintf("--cookie=%s", cookie))
}
```

Comment (pkalever) [reply internal/rbd/rbd_attach.go]: We need to defent further with a bool again, I would prefer keep it as it is just like others in the respective functions.

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Same as above.

**Review by pkalever — COMMENTED** [review]
Thanks @nixpanic

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: @nixpanic actually we don't have to check for empty `cookie` as it is passed from VolID which will never be empty?

**Comment by pkalever** [comment]
tests are failing:

```
[1mSTEP[0m: create a PVC and bind it to an app with normal user
Sep 30 06:45:48.183: INFO: Waiting up to &PersistentVolumeClaim{ObjectMeta:{rbd-1318  rbd-1318    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-rbd-sc,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},} to be in Bound state
Sep 30 06:45:48.183: INFO: waiting for PVC rbd-1318 (0 seconds elapsed)
Sep 30 06:45:50.192: INFO: waiting for PVC rbd-1318 (2 seconds elapsed)
Sep 30 06:45:50.199: INFO: Waiting for PV pvc-626fafad-908a-4a53-88de-cce70479d2b1 to bind to PVC rbd-1318
Sep 30 06:45:50.200: INFO: Waiting up to timeout=10m0s for PersistentVolumeClaims [rbd-1318] to have phase Bound
Sep 30 06:45:50.203: INFO: PersistentVolumeClaim rbd-1318 found and phase=Bound (3.129072ms)
Sep 30 06:45:50.203: INFO: Waiting up to 10m0s for PersistentVolume pvc-626fafad-908a-4a53-88de-cce70479d2b1 to have phase Bound
Sep 30 06:45:50.205: INFO: PersistentVolume pvc-626fafad-908a-4a53-88de-cce70479d2b1 found and phase=Bound (2.568213ms)
Sep 30 06:45:50.217: INFO: Waiting up to pod-run-as-non-root to be in Running state
Sep 30 06:55:50.229: FAIL: failed to validate normal user pvc and application binding with error timed out waiting for the condition
```
This doesn't look like an issue related to this PR

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.20/runs/2619/nodes/109/log/?start=0

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: no need to return bool from here? instead of that set `hasNBD` the same way you are handling the `hasNBDCookieSupport`variable

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
small suggestion. LGTM. requesting changes to hold this one till we have the next ceph release

@pkalever you also mention we have dependencies with ceph and kernel also. what if ceph has --cookie flag but the kernel does not support it. do you see any problem?

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Done now.

**Comment by pkalever** [comment]
> small suggestion. LGTM. requesting changes to hold this one till we have the next ceph release

Maybe you mean until the cookie PR is merged in ceph repo!?
`DNM` is already in place + there is a waiting task on the ceph PR which should keep this on track

> 
> @pkalever you also mention we have dependencies with ceph and kernel also. what if ceph has --cookie flag but the kernel does not support it. do you see any problem?

If the kernel doesn't support `cookie` then rbd-nbd (ceph) will just ignore it even if the user (ceph-csi) passes the `--cookie` option, hence it's safe and doesn't hurt to pass `--cookie` even if the kernel doesn't support it

File: internal/rbd/rbd_attach.go
Comment (nixpanic) [inline internal/rbd/rbd_attach.go:215]: if running the command fails, does it still make sense to set `hasNBD` to `true` as done above? Maybe set it after running the command and returning from the function on a failure?

Comment (pkalever) [reply internal/rbd/rbd_attach.go:215]: Sure :+1:

Comment (Madhu-1) [reply internal/rbd/rbd_attach.go:215]: isModuleLoaded should be set to false here also? am also thinking do we need an extra variable? can we just set `hasNBD` to false when running modproble or os.stat and `--help` fails?

File: internal/rbd/rbd_attach.go
Comment (Madhu-1) [inline internal/rbd/rbd_attach.go]: if os.Stat fails with any other error than IsNotExists `isModuleLoaded` remains `true`
`

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go:215]: I used hasNBD, later I thought the reviewers didn't like it.

Comment (Madhu-1) [reply internal/rbd/rbd_attach.go:215]: i think the request from Niels is to set `hasNBD` to false if running `--help` command fails

File: internal/rbd/rbd_attach.go
Comment (pkalever) [reply internal/rbd/rbd_attach.go]: Yes it should be set to true as before. We should just make sure /sys/module/nbd exists.

Comment (pkalever) [reply internal/rbd/rbd_attach.go]: > if os.Stat fails with any other error than IsNotExists isModuleLoaded remains true

Yes, this is intentional.

**Review by Madhu-1 — APPROVED** [review]

**Comment by pkalever** [comment]
@Mergifyio rebase

**Comment by pkalever** [comment]
@ceph/ceph-csi-contributors this PR is ready for review/merge. Thanks!

**Comment by pkalever** [comment]
@nixpanic could you also please help review this?

Thanks!

**Review by nixpanic — APPROVED** [review]

**Comment by pkalever** [comment]
```
X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

```

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.20/detail/mini-e2e_k8s-1.20/2998/pipeline

**Comment by pkalever** [comment]
```
X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute


```

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/upgrade-tests-cephfs/detail/upgrade-tests-cephfs/3579/pipeline

**Pull request merged**
